### PR TITLE
[Frontend][OpenMP] Allow implicit clauses to fail to apply

### PR DESCRIPTION
**Pull request by kparzysz**
The `linear(x)` clause implies `firstprivate(x)` on the compound construct if `x` is not an induction variable. With more construct combinations coming in OpenMP 6.0, the `firstprivate` clause may not be possible to apply, e.g. in "masked simd".
An additional benefit from this change is that it allows treating leaf constructs as combined constructs with a single constituent. Otherwise, a `linear` clause on a lone `simd` construct could imply a `firstprivate` clause that can't be applied.

### CODE REVIEW

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h [hunk 1]
```diff
@@ -1153,9 +1158,10 @@ template <typename C, typename H> bool ConstructDecompositionT<C, H>::split() {
   for (const ClauseTy *node : nodes) {
     if (skip(node))
       continue;
-    success =
-        success &&
+    bool result =
         std::visit([&](auto &&s) { return applyClause(s, node); }, node->u);
+    if (!isImplicit(node))
```
Comment (skatrak) [inline llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:1163]: Does this not have to be checked for `allocate` below?

**Review by skatrak — COMMENTED** [review]
It makes sense to me, but I think we should have a test to check it before merging.

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h (hunk 1 above)
Comment (kparzysz) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:1163]: `allocate` is not implied in 5.2 or 6.0 by any combination of directives and clauses.  I think it should be an error until some version of the spec allows it to be implied.

**Review by skatrak — APPROVED** [review]
Thank you, LGTM.

**Pull request merged**